src/agent.rs

```diff
@@ -66,69 +73,62 @@ impl Agent {
-            max_message_len,
+            max_sr_data_len,
         )?;
 
         Ok(Self {
             inner,
             data_recv: data_recv_mutex,
             handle,
-            max_message_len,
+            agent_thread,
+            max_sr_data_len,
             local_mr_recv,
             remote_mr_recv,
         })
     }
 
     /// Allocate a remote memory region
-    pub(crate) async fn request_remote_mr(&self, layout: Layout) -> io::Result<RemoteMemoryRegion> {
+    pub(crate) async fn request_remote_mr(&self, layout: Layout) -> io::Result<RemoteMr> {
         self.inner.request_remote_mr(layout).await
     }
 
-    /// Send a memory region metadata to the other side
-    pub(crate) async fn send_mr(&self, mr: Arc<dyn Any + Send + Sync>) -> io::Result<()> {
-        let request = if mr.is::<LocalMemoryRegion>() {
-            // Have checked the type in the if condition
-            #[allow(clippy::unwrap_used)]
-            let mr = mr.downcast::<LocalMemoryRegion>().unwrap();
-            let token = mr.token();
-            let ans = SendMRKind::Local(token);
-            if self.inner.mr_own.lock().await.insert(token, mr).is_some() {
-                return Err(io::Error::new(
-                    io::ErrorKind::InvalidInput,
-                    format!("the MR {:?} should be send multiple times", token),
-                ));
-            }
-            ans
-        } else {
-            let mr = mr.downcast::<RemoteMemoryRegion>().map_err(|m| {
-                io::Error::new(
-                    io::ErrorKind::InvalidData,
-                    format!(
-                        "this Mr {:?} can not be downcasted to RemoteMemoryRegion",
-                        m
-                    ),
-                )
-            })?;
-            SendMRKind::Remote(mr.token())
-        };
-        let request = Request {
-            request_id: AgentRequestId::new(),
-            kind: RequestKind::SendMR(SendMRRequest { kind: request }),
-        };
+    /// Send a local memory region metadata to the other side
+    pub(crate) async fn send_local_mr(&self, mr: LocalMr) -> io::Result<()> {
+        // Have checked the type in the if condition
+        let token = mr.token();
+        if self.inner.mr_own.lock().await.insert(token, mr).is_some() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("the MR {:?} should be send multiple times", token),
+            ));
+        }
+        let kind = SendMRKind::Local(token);
+        let request = RequestKind::SendMR(SendMRRequest { kind });
+        // this response is not important
+        let _ = self.inner.send_request(request).await?;
+        Ok(())
+    }
+
+    /// Send a remote memory region metadata to the other side
+    pub(crate) async fn send_remote_mr(&self, mr: RemoteMr) -> io::Result<()> {
+        let kind = SendMRKind::Remote(mr.token());
+        #[allow(clippy::mem_forget)]
+        mem::forget(mr);
```

Comment:
Comment the reason why we should skip the destructor of `mr`

Reply:
That was to prevent `double free`. But it's a bit of a memory waste, so I'm trying a better approach.

Reply:
Oh, This approach does not cause memory leaks. I read the doc and found that `mem::forget` skip the destructor and still reclaim the space taken by the variable `mr`. So this approach is good and I will comment the reason.